### PR TITLE
V0.0.4

### PR DESCRIPTION
**Pull request by stillonearth**
This PR includes improvements to high-level Rust wrappers to c bindings.

This example illustrates basic usage.

```rust
use mujoco_rust::{Model, Simulation};

const MJCF: &str = r#"<mujoco>
<worldbody>
    <light name="light0" diffuse=".5 .5 .5" pos="0 0 3" dir="0 0 -1"/>
    <geom name="geom0" type="plane" size="1 1 0.1" rgba=".9 0 0 1"/>
    <body name="body1" pos="0 0 1">
        <joint name="joint0" type="free"/>
        <geom name="geom1" type="box" size=".1 .2 .3" rgba="0 .9 0 1"/>
    </body>
</worldbody>
</mujoco>"#;

fn main() {
    let model = Model::from_xml_str(MJCF).unwrap();
    let simulation = Simulation::new(model);

    for _ in 0..1000 {
        println!(
            "time: {}, xpos: {:?}",
            simulation.state.time(),
            simulation.xpos()
        );
        simulation.step();
    }
}
```

* `mujoco_rust::Model` contains wrappers to `mj_Model` including initial positions, rotations and mesh data
* `mujoco_rust::Data` — `mj_Data`
* `mujoco_rust::Simulation` — is a new entity that's not present in MuJoCo, combining `mujoco_rust::Model` and `mujoco_rust::Data` to extract simulation data (`xpos`, `xquat`)

Wrappers extracted from `bevy_mujoco' crate

---

This should be done in one commit.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,25 @@
-# MuJoCo-rs
+# mujoco-rust
 MuJoCo bindings for Rust
 
 ## Installation requirements
-These bindings require that MuJoCo be installed before building.  The build system assumes that MuJoCo is installed into /usr/local on UNIX-based systems and C:\Program Files\MuJoCo on Windows, but any installation directory can be used, as long as either the environment variable MUJOCO_DIR or MUJOCO_PREFIX is set to the root of the installation.
\ No newline at end of file
+These bindings require that MuJoCo be installed before building.  The build system assumes that MuJoCo is installed into `/usr/.local` on UNIX-based systems and `C:\Program Files\MuJoCo` on Windows, but any installation directory can be used, as long as either the environment variable MUJOCO_DIR or MUJOCO_PREFIX is set to the root of the installation.
```
Comment (aaronlaw1215) [inline README.md]: /usr/.local ?

It should be /usr/local right?

File: mujoco/src/helpers.rs [hunk 2]
```diff
@@ -16,3 +17,126 @@ pub fn convert_err_buf(err_buf: Vec<u8>) -> String {
...
+    n_entries: usize,
+) -> Vec<f64>
+where
+    T: LocalFloat,
+{
+    let mut result_vec: Vec<f64> = Vec::new();
+
+    unsafe {
+        let entries = vec;
+        for i in 0..n_entries {
```
Comment (aaronlaw1215) [inline mujoco/src/helpers.rs]: How often is this called?  Are there any ways of flattening the doubly-nested loop here?

File: mujoco/src/helpers.rs [hunk 3]
```diff
@@ -16,3 +17,126 @@ pub fn convert_err_buf(err_buf: Vec<u8>) -> String {
...
+                result_vec.push((*entries.add(i * element_size + j)).to_f32());
+            }
+        }
+    }
+
+    result_vec
+}
+
+/// Get String from an pointer to a null-terminated char array pointer
+pub(crate) fn extract_string(array: *mut i8) -> String {
```
Comment (aaronlaw1215) [inline mujoco/src/helpers.rs]: Would CStr not do exactly this?

`Cstr::from_ptr(array).to_str().to_owned()`

File: mujoco/src/sim.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..self.model.nu() {
+            unsafe {
+                *raw_vec.add(i) = control[i];
+            }
+        }
+    }
+
+    /// Evaulate constraint forces ans sensors
```
Comment (aaronlaw1215) [inline mujoco/src/sim.rs]: typo: "ans" should be "and"

File: mujoco/src/sim.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        xpos
+    }
+
+    /// Returns rotations of bodies
+    pub fn xquat(&self) -> Vec<[f64; 4]> {
+        let mj_data = self.state.ptr();
+        let raw_vec = unsafe { (*mj_data).xquat };
+        let raw_quat =
```
Comment (aaronlaw1215) [inline mujoco/src/sim.rs]: This is a bit of a nit, but wouldn't it be better for readability to use a proper Quaternion type?

File: mujoco/src/state.rs [hunk 6]
```diff
@@ -1,75 +1,57 @@
...
 
 /// The time-dependent state of a MuJoCo simulation. Analagous to mjData from
 /// the C API
 #[derive(Debug)]
-pub struct State<'a> {
-    ptr: *mut mujoco_rs_sys::no_render::mjData,
-    // TODO: Perhaps this shouldn't be here, and we should manage cloning and resetting
-    // in the `Sim` struct
-    model: &'a Model,
+pub struct State {
```
Comment (aaronlaw1215) [inline mujoco/src/state.rs:6]: This is better than holding on to the Model reference.  Good job :smile:

**Review by aaronlaw1215 — COMMENTED** [review]
Overall this is good.

I do have a few comments that need to be addressed before I merge though.  And I have a general comment regarding generated files: don't check in generated files.  The files "no_render.rs" and "render.rs" in the generated folder should not be checked in.  Please check your .gitignore.  I thought I added an entry for the entire "generated" folder.

File: README.md (hunk 1 above)
Comment (stillonearth) [reply README.md]: It's actually `~/.local/mujoco`

File: mujoco/src/sim.rs (hunk 5 above)
Comment (stillonearth) [reply mujoco/src/sim.rs]: Which quat implementation? 

https://github.com/PistonDevelopers/quaternion — small footprint
or perhaps https://docs.rs/nalgebra/latest/nalgebra/?search=Quaternion

**Comment by stillonearth** [comment]
- nalgebra `Vec3`, `Quaternion` where appropriate
- fixes to upmentioned issues

**Comment by stillonearth** [comment]
- cleaned up history

File: README.md (hunk 1 above)
Comment (aaronlaw1215) [reply README.md]: That makes sense.

File: mujoco/src/sim.rs (hunk 5 above)
Comment (aaronlaw1215) [reply mujoco/src/sim.rs]: I was thinking nalgebra's, but really any is fine.

**Review by aaronlaw1215 — APPROVED** [review]

**Comment by stillonearth** [comment]
Great! Can you bump version on crates.io?

**Pull request merged**